### PR TITLE
explore/12-tcnn-abilities 

### LINKED ISSUE #12: check TCNN abilitiies
to do:
- reimplement TCNN
- check its ability to avoid regressions
- try a neighbor prioritization approach during local search using TCNN prediction sorting

### PR DESCRIPTION
**Pull request by zinchse**
**tldr:** TCNN is able to predict query execution time by its plan (at least with simple memoization), but its ability to work well on new plans/queries is questionable.

# Plan Representation

A special structure for processing binary trees has been implemented, mainly through the use of two arguments - `vertices` and `edges`. The `vertices` are simply a set of vector representations of the plan nodes that were obtained by simple vectorisation during tree traversal. The `edges`, on the other hand, are triples of indices for each node that point to its *neighbourhood*. A neighbourhood is the node itself, and all its children (in our case, at most two). To simplify the tree convolution, all missing children were replaced by a special vertex-padding.

# Data Preprocessing

## Timeout

By simply removing plans with `TIMEOUT`s, we deprive the NN of information about frankly bad parameter sets, which makes the probability that it will estimate them correctly drop significantly. For this purpose, each such plan was labelled with the `TIMEOUT` value used at that moment. This solution is also not ideal, as it led to underestimation of execution time in some cases. However, no other ideas for solutions to this problem were found. 

## Weighted Binary Tree Dataset

In order to spare the neural network from the complexities associated with plan *collisions* and ambiguous labels, it was decided to *group* all plans with the group assigned a label in the form of an *average* execution time. This technique at the same time allowed us to **speed up training** significantly, since there are far fewer unique plans. However, to avoid bias in the NN predictions, each unique plan was assigned a weight (`frequency`) equal to the number of its duplicates.

## Dataloader

for the possibility of efficient processing of plans by batches it was necessary to properly organise the order of channels in vertices, allowing to reuse already efficiently written implementations of the same-name convolutions. Also, to make the predictions independent of the batches, we had to hardcode the size of the paddings for the tree as the maximum tree size. This will need to be given special attention in a production application - by design, convolution layers on trees can only handle a batch of trees of the same size.

# Neural Network

## The role of Binary Tree Convolution

image

It has been experimentally shown that binary tree convolutional NN actually **help to extract a more informative representation of trees**, thus simplifying the prediction task.

## Layers

For convenience, special layers with `torch`-like interface were implemented to allow processing of objects represented by `vertices` and `edges`. All such layers can also be conveniently stacked using a special module `BinaryTreeSequential`. In the end, the architecture represents two blocks - a binary tree convolutional block on binary trees, which extracts information in the form of some vector, and then a block from a fully connected NN.

## Features
image

Among the key features compared to existing implementations are a) the availability of tests, and b) the real implementation of normalisation layers. It was found that `InstanceNormalisation` works better than the others.

# Ability to generalise

All experiments were conducted over queries for `imdb` database (`JOB` and `sample_queries` benchmarks).

## Train Data

image

We were able to select an architecture that achieves near 0 error on the training dataset. However, by its structure, the dependence under study is quite complex, and even the trained NN continued to make errors in special cases. Thus, for example, `TIMEOUT` processing led to underestimation of the real time of plan execution, and the NN, relying on similar plans, quite naturally predicted higher value than the label. Another difficulty is a situation similar to collision - when two very similar plans have completely different labels, as a result of which the approximated surface becomes too sharp and its complete modelling is difficult. 

Moreover, it has even been shown separately that this situation *prevents the convolutional layer from producing an efficient representation* - if we replace the tree representation with **random&& noise, the learning dynamics will be even **faster**! 

The final layer in the chosen architecture (`softplus`) also played its role, which made it difficult to learn weights in regions with small predictions. However, the use of such a layer allowed us to effectively solve the problem with the limitation on the positivity of predictions.

## Validation Data

image

The error value on the validation dataset is significantly higher than on the training dataset.  The reason for this, probably, is the complex geometry of the studied dependence and *unpredictable* sharp elevations.

## Test Data

The test dataset consisted of plans from `sample_queries`, whose structure was also present in `JOB`. Surprisingly, the test performance turned out to be **even better** than the validation one. Likely, the reason for this was the similarity of the plans - the probability that a randomly chosen subtree of size up to 4 would be found in the training dataset was much higher in the test set compared to the validation set.

| Subtree Height | Dataset | Match Probability   |
|--------|-----------|---------------------|
| 1      | train     | 0.853 ± 0.005       |
| 2      | train     | 0.723 ± 0.006       |
| 3      | train     | 0.621 ± 0.006       |
| 4      | train     | 0.527 ± 0.005       |
| 1      | test      | 0.975 ± 0.006       |
| 2      | test      | 0.952 ± 0.009       |
| 3      | test      | 0.933 ± 0.011       |
| 4      | test      | 0.914 ± 0.013       |

# Insights

Experiments using noise instead of real representations not only revealed that the studied surface is complicated by the problem of collisions but also highlighted that **statistics are a more important feature than the tree structure**. Indeed, by utilizing only statistics, the NN effectively gains access to information about the expected amount of data to be processed. In this case, it only needs to adjust the appropriate cost coefficients corresponding to the time scale. With the additional information about the tree structure, the network can model **contextual dependencies** — depending on the observed subtree, represented by the information about the operations used, the studied cost coefficients can vary, thereby **correcting** the errors of cardinality models, and so on.

This raises an important question: *"Are the model's predictions on new plans more useful than those of the cost model?"* To answer this question, it is possible to may be useful to shift from a regression task to a plan ranking task, using the `cost model` as a baseline.

### CODE REVIEW

**Comment by zinchse** [comment]
next steps (i'll create issue for them soon): 
- add costs in plan representation;
- compare best nn with cost model on ranking problem.
